fmpz/mul_2exp.c

```diff
@@ -9,38 +10,103 @@
     (at your option) any later version.  See <https://www.gnu.org/licenses/>.
 */
 
-#include <gmp.h>
-#include "flint.h"
-#include "ulong_extras.h"
 #include "fmpz.h"
+#include <string.h>
 
 void
 fmpz_mul_2exp(fmpz_t f, const fmpz_t g, ulong exp)
 {
-    fmpz d = *g;
+    slong c1 = *g;
+    ulong c1abs, c1bits;
+    __mpz_struct * mf;
 
-    if (!COEFF_IS_MPZ(d))       /* g is small */
+    if (c1 == 0)
     {
-        ulong dabs = FLINT_ABS(d);
-        ulong bits = FLINT_BIT_COUNT(dabs);
-        if (bits == 0)
+        fmpz_zero(f);
+        return;
+    }
+
+    c1abs = FLINT_ABS(c1);
+    c1bits = FLINT_BIT_COUNT(c1abs);
+
+    if (c1bits + exp <= FLINT_BITS - 2)  /* Result fits inside a small fmpz */
+    {
+        if (COEFF_IS_MPZ(*f))
+            _fmpz_clear_mpz(*f);
+
+        *f = c1 << exp;
+    }
+    else if (c1bits <= FLINT_BITS - 2)   /* g is small */
+    {
+        ulong expred = exp % GMP_NUMB_BITS;
+        int alloc = 1 + (FLINT_BITS != GMP_NUMB_BITS) + exp / GMP_NUMB_BITS
+            + ((c1bits + expred) > GMP_NUMB_BITS);
+        mp_limb_t * limbs;
+        slong tmp;
+
+        /* Ensure enough limbs are allocated for f */
+        if (!COEFF_IS_MPZ(*f))
         {
-            fmpz_set_si(f, 0);
+            /* TODO: Initialize the new mpz with alloc limbs instead of
+             * reallocating them. */
+            mf = _fmpz_new_mpz();
+            *f = PTR_TO_COEFF(mf);
+            _mpz_realloc(mf, alloc);
         }
-        else if (bits + exp <= FLINT_BITS - 2)  /* result will fit in small */
+        else
         {
-            fmpz_set_si(f, d << exp);
+            mf = COEFF_TO_PTR(*f);
+            if (mf->_mp_alloc < alloc)
+                _mpz_realloc(mf, alloc);
         }
-        else                    /* result is large */
+        limbs = mf->_mp_d;
+        mf->_mp_size = (c1 > 0) ? alloc : -alloc;
+        memset(limbs, 0, sizeof(mp_limb_t) * (alloc - 1));
+
+        tmp = (slong) (c1bits + expred) - GMP_NUMB_BITS;
+        if (tmp <= 0)
         {
-            __mpz_struct * mf = _fmpz_promote(f);   /* g is saved */
-            flint_mpz_set_si(mf, d);
-            mpz_mul_2exp(mf, mf, exp);
+            limbs[alloc - 1] = c1abs << expred;
+#if FLINT_BITS != GMP_NUMB_BITS
+            return;
+#endif
         }
+#if FLINT_BITS == GMP_NUMB_BITS
+        else
+        {
+            limbs[alloc - 1] = c1abs >> (FLINT_BITS - expred);
+            limbs[alloc - 2] = c1abs << expred;
+        }
+#else   /* FLINT_BITS == 2 * GMP_NUMB_BITS */
```

Comment:
We do not support nails, so this distinction is not necessary. GMP_NUMB_BITS is always FLINT_BITS.

Reply:
I thought they would be an array of 32-bit integers in the case for 64-bit Windows, but do they actually leave 32-bit spaces in between the limbs? Sorry for my lack of knowledge.